source/io/module_eggShell.js

```diff
@@ -361,21 +386,91 @@
                 throw new Error('Empty viaText provided, nothing to run');
             }
 
-            if (userPrintFunction === undefined) {
-                console.warn('Failing to call eggShell.setPrintFunction prior to eggShell.loadVia may result in missed messages');
-            }
-
             var viaTextLength = Module.lengthBytesUTF8(viaText);
             var viaTextPointer = Module.coreHelpers.writeJSStringToHeap(viaText);
+
+            var printText = '',
+                printTextErr = '';
+            var origPrint = Module.print,
+                origPrintErr = Module.printErr;
+
+            Module.print = function (text) {
+                printText += text + '\n';
+                origPrint(text);
+            };
+
+            Module.printErr = function (textErr) {
+                printTextErr += textErr + '\n';
+                origPrintErr(textErr);
+            };
+
             var result = Module._EggShell_REPL(v_userShell, viaTextPointer, viaTextLength);
             Module._free(viaTextPointer);
-            return result;
+            Module.print = origPrint;
+            Module.printErr = origPrintErr;
+
+            if (result !== 0) {
+                throw new Error('Loading VIA failed for the following reason: ' + niErrorEnum[result] +
+                    ' (error code: ' + result + ')' +
+                    ' (stdout: ' + printText + ')' +
+                    ' (stderr: ' + printTextErr + ')');
+            }
         };
 
-        Module.eggShell.executeSlicesUntilWait = publicAPI.eggShell.executeSlicesUntilWait = function (slices) {
-            // returns < 0 if should be called again ASAP, 0 if nothing to run, or positive value N if okay
-            // to delay up to N milliseconds before calling again
-            return EggShell_ExecuteSlices(v_userShell, slices);
+        // executeSlicesUntilWait
+        // numSlices (optional): The minimum number of slice sets to run before checking if maxTimeMS has passed.
+        //    The larger the value the less overhead for execution and the quicker the diagram progresses
+        //    One slice set corresponds to 10 slices and at a minimum one slice set executes per invocation
+        // millisecondsToRun (optional): The amount of time in milliseconds vireo can execute slice sets before vireo saves state and returns
+        // return value (type ExecSlicesResult):
+        //     returns < 0 if should be called again ASAP, 0 if nothing to run, or positive value N if okay
+        //     to delay up to N milliseconds before calling again
+        Module.eggShell.executeSlicesUntilWait = publicAPI.eggShell.executeSlicesUntilWait = function (numSlices, millisecondsToRun) {
```

Comment:
If the "numSlices" and "millisecondsToRun" are optional, do we not need to check whether the values are valid and if not use pre-defined default values before calling into EggShell_ExecuteSlices()?

Reply:
We don't "need" to because asm.js does a  `<number> | 0` to coerce it to Int32 semantics in the js runtime.

You can see it in the generated output at the _EggShell_ExecuteSlices function in https://unpkg.com/vireo@7.0.0/dist/asmjs-unknown-emscripten/debug/vireo.js

reproduced below (this is an older version that took only two parameters while it now takes 3):
```
function _EggShell_ExecuteSlices($0, $1) {
 $0 = $0 | 0; //@line 11101
 $1 = $1 | 0; //@line 11102
 var $2 = 0, $3 = 0, $4 = 0, $5 = 0, $6 = 0, $7 = 0, $8 = 0, $9 = 0, label = 0, sp = 0; //@line 11103
 sp = STACKTOP; //@line 11104
 STACKTOP = STACKTOP + 16 | 0; //@line 11105
 // ...
}
```

And the value of the arguments if not provided is `undefined` and the value of `undefined | 0` is the numeric value `0`. The behavior of not passing an argument is subtly described in the comments as `at a minimum one slice set executes per invocation`.

I agree that it is a subtle behavior so I'm not opposed to putting the checks in. My only concern was that executeSlices is on a fast path and I want to reduce overhead on a fast path but it could definitely be a premature optimization. It is very likely run times will optimize the path anyway and the type checks may even assist that.